polygeist/tools/cgeist/Lib/CGExpr.cc

```diff
@@ -248,30 +256,33 @@ mlir::Attribute MLIRScanner::InitializeValueByInitListExpr(mlir::Value ToInit,
 
         mlir::Value Next;
         if (auto MT = ToInit.getType().dyn_cast<MemRefType>()) {
-          auto Shape = std::vector<int64_t>(MT.getShape());
+          llvm::SmallVector<int64_t> Shape(MT.getShape());
           assert(!Shape.empty());
           Shape[0] = ShapedType::kDynamic;
 
-          if (MT.getElementType()
-                  .isa<mlir::sycl::AccessorType,
-                       mlir::sycl::AccessorImplDeviceType,
-                       mlir::sycl::ArrayType, mlir::sycl::ItemType,
-                       mlir::sycl::NdItemType, mlir::sycl::GroupType,
-                       mlir::sycl::VecType>()) {
-            llvm_unreachable("not implemented yet");
-          }
-
           mlir::Type ET;
           if (auto ST =
                   MT.getElementType().dyn_cast<mlir::LLVM::LLVMStructType>()) {
             ET = mlir::MemRefType::get(Shape, ST.getBody()[I],
                                        MemRefLayoutAttrInterface(),
                                        MT.getMemorySpace());
-          } else if (auto ST = MT.getElementType()
-                                   .dyn_cast<mlir::sycl::ItemBaseType>()) {
-            ET = mlir::MemRefType::get(Shape, ST.getBody()[I],
-                                       MemRefLayoutAttrInterface(),
-                                       MT.getMemorySpace());
+          } else if (sycl::isSYCLType(MT.getElementType())) {
+            mlir::Type ElemTy = MT.getElementType();
+            ET =
+                TypeSwitch<mlir::Type, mlir::MemRefType>(ElemTy)
+                    .Case<mlir::sycl::IDType, mlir::sycl::RangeType>(
+                        [&](auto T) {
+                          return mlir::MemRefType::get(
+                              Shape, MT.getElementType(),
+                              MemRefLayoutAttrInterface(), MT.getMemorySpace());
+                        })
+                    .Case<mlir::sycl::ItemBaseType>([&](auto ElemTy) {
```

Comment:
ElemTy -> Ty (given there is already an ElemTy at outer scope). I know the local `ElemTy` hides the outer one, but still...